design-proposals/klip-43-timestamp-data-type-support.md

```diff
@@ -0,0 +1,206 @@
+
+## What is in scope
+
+* Add TIMESTAMP type to KSQL
+* Support TIMESTAMP arithmetic and comparisons
+* Allow window units (HOUR, DAY etc) to be used in timestamp functions
+* Support TIMESTAMP usage in STRUCT, MAP and ARRAY
+* Serialization and de-serialization of TIMESTAMPs to Avro, JSON, Protobuf and Delimited formats
+* Update existing built-in functions to use the TIMESTAMP data type
+* Casting TIMESTAMP to and from BIGINT and STRING
+
+## What is not in scope
+* Changing the ROWTIME data type. We will eventually want this to happen, but that is a separate
+discussion.
+* Support for dates before Unix epoch - this is not supported by Kafka. There is a [KIP](https://cwiki.apache.org/confluence/display/KAFKA/KIP-228+Negative+record+timestamp+support)
+for negative timestamps, but until that is implemented, KSQL can only support positive timestamps.
+* DATE and TIME types - because TIMESTAMPs represent a point in time, DATE and TIME types
+would be useful if a user wants to represent a less specific time. These would be added after
+TIMESTAMP is implemented though.
+
+## Public APIS
+
+The TIMESTAMP data type will store a point in time after Unix epoch without timezone information.
+The syntax is as follows:
+
+```roomsql
+CREATE STREAM stream_name (time TIMESTAMP, COL2 STRING) AS ...
+CREATE TABLE table_name (col1 STRUCT<field TIMESTAMP>) AS ...
+```
+
+TIMESTAMPS will be displayed in console as strings in ODBC canonical format with millisecond precision:
+
+```roomsql
+> SELECT time FROM stream_name EMIT CHANGES;
+
++------------------------+
+|time                    |
++------------------------+
+|1994-11-05 13:15:30:112 |
+```
+
+TIMESTAMPS can be represented by milliseconds from Unix epoch or by date strings:
+
+```roomsql
+INSERT INTO stream_name VALUES (1605927509166);
+INSERT INTO stream_name VALUES ("1994-11-05 13:15:30");
+```
+
+## Design
+
+### Serialization/Deserialization
+
+TIMESTAMPs will be handled by the `java.time.Instant` class in UTC within KSQL. The corresponding Kafka Connect type is
+[org.apache.kafka.connect.data.Timestamp](https://kafka.apache.org/0100/javadoc/org/apache/kafka/connect/data/Timestamp.html).
+They are represented as long types in Schema Registry, but also come with a tag indicating that it
+is a timestamp, so they should be distinguishable from long types when handling serialized values in KSQL.
+
+#### Avro
+
+Avro schemas represent timestamps as
+```json
+{
+  "type": "long",
+  "logicalType": "timestamp-millis"
+}
+```
+The Avro deserializer KSQL uses supports this.
+
+#### Protobuf
+Protobuf 3 has a Timestamp type. The Protouf deserializer KSQL uses supports this.
+
+#### JSON/Delimited
+
+Timestamps will get stored in JSON and CSV files as long values. The KSQL JSON and delimited deserializers
+will be updated to parse timestamps.
+
+### UDFs
+
+The following UDFs should be updated to use the TIMESTAMP type instead of BIGINT:
+
+* TIMESTAMPTOSTRING
+* STRINGTOTIMESTAMP
```

Comment:
This is gonna cause some confusion between millis and seconds. We really need to discuss this with the rest of the team, how should we handle timestamps, as millis or seconds? If seconds, then we don't want to break compatibility with TIMESTAMPTOSTRING and STRINGTOTIMESTAMP, which currently support milliseconds.

If we do milliseconds, then we're good with the above UDF.
If we do seconds, we perhaps want to add a `to_timestamp(string, format)` function that returns the UNIX epoch in seconds. `to_timestamp` is used in other DB too in the same format. Postgres adds `to_timestamp(double)` which converts seconds to timestamp.